relays/ethereum/src/exchange_loop.rs

```diff
@@ -0,0 +1,267 @@
+		TransactionProofsRelayState {
+			best_processed_header_number: self.best_processed_header_number,
+		}
+	}
+
+	fn set_state(&mut self, state: &TransactionProofsRelayState<BlockNumber>) {
+		self.best_processed_header_number = state.best_processed_header_number;
+	}
+}
+
+/// Run proofs synchronization.
+pub fn run<P: TransactionProofPipeline>(
+	mut storage: impl TransactionProofsRelayStorage<BlockNumber = BlockNumberOf<P>>,
+	source_client: impl SourceClient<P>,
+	target_client: impl TargetClient<P>,
+	exit_signal: impl Future<Output = ()>,
+) {
+	let mut local_pool = futures::executor::LocalPool::new();
+
+	local_pool.run_until(async move {
+		let mut retry_backoff = retry_backoff();
+		let mut state = storage.state();
+		let mut current_finalized_block = None;
+
+		let exit_signal = exit_signal.fuse();
+
+		futures::pin_mut!(exit_signal);
+
+		loop {
+			let iteration_result = run_loop_iteration(
+				&mut storage,
+				&source_client,
+				&target_client,
+				&mut state,
+				&mut current_finalized_block,
+			)
+			.await;
+
+			match iteration_result {
+				Ok(_) => {
+					retry_backoff.reset();
+
+					select! {
+						_ = source_client.tick().fuse() => {},
+						_ = exit_signal => return,
+					}
+				}
+				Err(_) => {
+					let retry_timeout = retry_backoff.next_backoff().unwrap_or(CONNECTION_ERROR_DELAY);
+
+					select! {
+						_ = async_std::task::sleep(retry_timeout).fuse() => {},
+						_ = exit_signal => return,
+					}
+				}
+			}
+		}
+	});
+}
+
+/// Run exchange loop until we need to break.
+async fn run_loop_iteration<P: TransactionProofPipeline>(
+	storage: &mut impl TransactionProofsRelayStorage<BlockNumber = BlockNumberOf<P>>,
+	source_client: &impl SourceClient<P>,
+	target_client: &impl TargetClient<P>,
+	state: &mut TransactionProofsRelayState<BlockNumberOf<P>>,
+	current_finalized_block: &mut Option<(P::Block, RelayedBlockTransactions)>,
+) -> Result<(), ()> {
+	let best_finalized_header_id = match target_client.best_finalized_header_id().await {
+		Ok(best_finalized_header_id) => {
+			log::trace!(
+				target: "bridge",
+				"Got best finalized {} block from {} node: {:?}",
+				P::SOURCE_NAME,
+				P::TARGET_NAME,
+				best_finalized_header_id,
+			);
+
+			best_finalized_header_id
+		}
+		Err(err) => {
+			log::error!(
+				target: "bridge",
+				"Failed to retrieve best {} header id from {} node: {:?}. Going to retry...",
+				P::SOURCE_NAME,
+				P::TARGET_NAME,
+				err,
+			);
+
+			return Err(());
+		}
+	};
+
+	loop {
+		// if we already have some finalized block body, try to relay its transactions
+		if let Some((block, relayed_transactions)) = current_finalized_block.take() {
+			let result = relay_block_transactions(source_client, target_client, &block, relayed_transactions).await;
+
+			match result {
+				Ok(relayed_transactions) => {
+					log::info!(
+						target: "bridge",
+						"Relay has processed {} block #{}. Total/Relayed/Failed transactions: {}/{}/{}",
```

Comment:
```suggestion
						"Relay has processed {} block #{}. Total: {}, Relayed: {}, Failed: {} Transactions",
```

I think this is easier to read at a glance, but it's a bit of a nitpick